docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,725 @@
+config. This will also apply on restarting a stalled workflow that timed out.
+
+We should also improve the visibility of stalled workflows and incomplete tasks
+  - `cylc scan` and the UI gscan panel should clearly indicate if a running
+  workflow is stalled or has incomplete tasks
+  - similarly for stopped workflows (also: "not started", "ran to completion",
+    "died", etc.).
+  - we need a mechanism to ensure that any workflows that stopped unexpectedly
+    don't get missed and that users can easily find their most recent workflows
+    (e.g. `cylc review` orders by last activity).
+
+## New Event Handler
+
+We should add a new **incomplete task** event that is triggered whenever a task
+finishes without completing all of its required outputs.
+
+## Runahead limit
+
+See https://github.com/cylc/cylc-flow/issues/4258. Some users understand
+runahead limiting as preventing the fastest tasks from getting too far ahead of
+the slowest.
+
+Incomplete tasks and partially satisfied prerequisites both represent tasks
+that will, or at least might after user intervention, still run at their cycle
+point, so they should count towards the runahead limit.
+
+(They should not count toward an *active cycle point* based limit, if we decide
+to implement that).
+
+Note that future triggers have the potential to cause complications with the
+runahead limit by spawning tasks at old cycles. This kind of issue would have
+been obvious at Cylc 7 because the task waiting for the future trigger would
+already have been in the task pool.
+
+## Cylc 7 compatibility
+
+Cylc 7 workflows will not be using the new syntax which means they will only
+have required outputs (but that's OK - Cylc 7 outputs **are** "required" as
+explained above).
+
+They also use suicide triggers to remove:
+- pre-spawned waiting tasks that would never run (i.e. from unused alternate
+  branches)
+- failed tasks that don't need to be retained in the pool
+
+For example, in this Cylc 8 graph the `c1 => c2` branch does not get spawned
+at all unless `a` fails:
+```
+a? => b1 => b2
+a:fail? => c1 => c2
+b2 | c2 => d
+```
+
+The Cylc 7 version of this already works out of the box on Cylc 8 master, for
+backward compatibility:
+```
+a => b1 => b2
+a:fail => c1 => c2
+b2 | c2 => d
+a => !c1 & !c2
+a:fail => !a & !b1 & !b2
+d => !a # prevent "a" disappearing from the GUI too quickly
+```
+
+This works because Cylc 8 suicide triggers spawn (or update) children just like
+normal triggers. Once all of a task's suicide prerequisites are satisfied it
+is removed immediately. A task that has a single suicide trigger and would not
+otherwise be spawned in Cylc 8 will have a transient presence in the task pool
+because the suicide trigger that spawns it also satisfies its only suicide
+prerequisite.
+
+It doesn't matter if a task still has unsatisfied normal prerequisites when it
+suicides. The other triggers can't spawn it again in the same flow thanks to
+our conditional reflow prevention mechanism.
```

Comment:
That's clever. I hope we retain this behaviour if we make any changes to how suicide triggers are handled.

Reply:
Not planning to change this at all :+1:

Reply:
"Once all of a task's suicide prerequisites are satisfied it is removed immediately"
Presumably not if it is in the `preparing`, `submitted` or `running` state?
In this case I assume we need to prevent further tasks being spawned (remove it from all flows?) and log a warning (since it doesn't make sense for a task in these states to be suicided).

Reply:
>  Presumably not if it is in the preparing, submitted or running state?

Yes, we remove it immediately even then.  Even in Cylc 7 we do that actually.  It doesn't cause any problems. You'll get a warning about suiciding an active task.  Task messages will be logged and ignored, from the orphans.